util/dvsim/SimCfg.py

```diff
@@ -539,20 +543,67 @@ def _gen_results(self, run_results):
         is enabled, then the summary coverage report is also generated. The final
         result is in markdown format.
         '''
+        def indent_by(level):
+            return " " * (4 * level)
+
         def create_failure_message(test, line, context):
-            spaces = " " * 12
-            message = [f"    * {test.qual_name}", ""]
+            message = [f"{indent_by(2)}* {test.qual_name}\\"]
             if line:
                 message.append(
-                    f"{spaces}Line {line}, in log {test.get_log_path()}")
+                    f"{indent_by(2)}  Line {line}, in log " +
+                    test.get_log_path())
             else:
-                message.append(f"{spaces}Log {test.get_log_path()}")
+                message.append(f"{indent_by(2)} Log {test.get_log_path()}")
             if context:
-                lines = [f"{spaces}{c.rstrip()}" for c in context]
+                message.append("")
+                lines = [f"{indent_by(4)}{c.rstrip()}" for c in context]
                 message.extend(lines)
             message.append("")
             return message
 
+        def create_bucket_report(buckets):
+            """Creates a report based on the given buckets.
+
+            The buckets are sorted by descending number of failures. Within
+            buckets this also group tests by unqualified name, and just a few
+            failures are shown per unqualified name.
+
+            Args:
+              buckets: A dictionary by bucket containing triples
+                (test, line, context).
+
+            Returns:
+              A list of text lines for the report.
+            """
+            by_tests = sorted(buckets.items(),
+                              key=lambda i: len(i[1]),
+                              reverse=True)
+            fail_msgs = ["\n## Failure Buckets", ""]
+            for bucket, tests in by_tests:
+                fail_msgs.append(f"* `{bucket}` has {len(tests)} failures:")
+                unqual_dict = collections.defaultdict(list)
+                for (test, line, context) in tests:
+                    unqual_dict[test.name].append((test, line, context))
+                for test_count, (name, qual_tests) in enumerate(
+                        unqual_dict.items()):
+                    if test_count == _MAX_TESTS:
+                        fail_msgs.append(
+                            f"{indent_by(1)}* ... and "
+                            f"{len(unqual_dict) - test_count} more tests.")
+                        break
+                    fail_msgs.append(f"{indent_by(1)}* Test {name} has "
+                                     f"{len(qual_tests)} failures.")
+                    for count, (test, line, context) in enumerate(qual_tests):
```

Comment:
The `qual_tests` is a list. I think we can probably just slice the array to limit what's printed. That way we can avoid the looping construct and having check each iteration. 

Come to think of it  - perhaps the earlier for loop could follow the same.

Reply:
More pythonic, thanks. Notice the top dict.items() needs to be cast to a list for slicing.